lib/vrl/DESIGN.md

```diff
@@ -0,0 +1,219 @@
+# Vector Remap Language: Guiding Design Principles
+
+This document describes the high-level goals and directions of the Vector Remap
+Language (VRL). It is intended to help guide discussions and decisions made
+during the development of VRL itself. The document is less relevant to _users_
+of VRL, the [language documentation][docs] exists for that audience.
+
+[docs]: https://vrl.dev
+
+## Table of Contents
+
+* [The Zen of VRL](#the-zen-of-vrl)
+* [Why VRL](#why-vrl)
+* [Target Audience](#target-audience)
+* [Language Limits](#language-limits)
+* [Conventions](#conventions)
+  * [Functions](#functions)
+    * [Composition](#composition)
+    * [Naming](#naming)
+    * [Return Types](#return-types)
+    * [Mutability](#mutability)
+    * [Fallibility](#fallibility)
+    * [Signatures](#signatures)
+  * [Errors](#errors)
+* [Patterns](#patterns)
+  * [Error Chaining](#error-chaining)
+
+## The Zen of VRL
+
+In the spirit of [The Zen of Python][PEP20].
+
+- Beautiful is better than ugly.
+- Explicit is better than implicit.
+- Simple is better than complex.
+- Sparse is better than dense.
+- Readability trumps writing convenience.
+- Special cases aren't special enough to break the rules.
+- Although practicality beats purity.
+- Function calls over syntax.
+- Errors may never occur at run-time.
+- Unless explicitly marked as such.
+- There should be one — and preferably only one — obvious way to do it.
+- Performance matters.
+- Provide observability remapping solutions, nothing else.
+
+[PEP20]: https://www.python.org/dev/peps/pep-0020/
+
+## Why VRL
+
+VRL exists to solve the need for a _simple_ and _performant_ **domain specific
+language** (DSL) to remap observability data.
+
+Its purpose is to hit a sweet spot between Turing complete languages such as
+Lua, and static transforms such as Vector's `rename_fields`. It should be
+flexible enough to cover most remapping use-cases, without requiring operators
+to write complex scripts that are difficult to reason about and incur
+a significant performance penalty.
```

Comment:
```suggestion
Its purpose is to hit a sweet spot between Turing complete languages such as
Lua, and static transforms such as Vector's old `rename_fields`. It should be
flexible enough to cover most remapping use-cases, without requiring the
flexibility and downsides of a full programming language, such as poor readability
and performance.
```

I think these changes a little more specific with the goals. Take it or leave it.